projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs

```diff
@@ -63,339 +60,233 @@ internal sealed class AutorecoveringConnection : IConnection
 
         public AmqpTcpEndpoint[] KnownHosts
         {
-            get { return m_delegate.KnownHosts; }
-            set { m_delegate.KnownHosts = value; }
+            get => m_delegate.KnownHosts;
+            set => m_delegate.KnownHosts = value;
         }
 
-        public int LocalPort
-        {
-            get { return m_delegate.LocalPort; }
-        }
+        public int LocalPort => m_delegate.LocalPort;
 
-        public ProtocolBase Protocol
-        {
-            get { return m_delegate.Protocol; }
-        }
+        public ProtocolBase Protocol => m_delegate.Protocol;
 
-        public IDictionary<string, RecordedExchange> RecordedExchanges
-        {
-            get { return m_recordedExchanges; }
-        }
+        public IDictionary<string, RecordedExchange> RecordedExchanges => m_recordedExchanges;
+
+        public IDictionary<string, RecordedQueue> RecordedQueues => m_recordedQueues;
+
+        public int RemotePort => m_delegate.RemotePort;
+
+        public IDictionary<string, object> ServerProperties => m_delegate.ServerProperties;
+
+        public IList<ShutdownReportEntry> ShutdownReport => m_delegate.ShutdownReport;
 
-        public IDictionary<string, RecordedQueue> RecordedQueues
+        IProtocol IConnection.Protocol => Endpoint.Protocol;
+        
+        private enum RecoveryCommand
         {
-            get { return m_recordedQueues; }
+            /// <summary>
+            /// Transition to auto-recovery state if not already in that state.
+            /// </summary>
+            BeginAutomaticRecovery,
+            /// <summary>
+            /// Attempt to recover connection. If connection is recovered, return
+            /// to connected state.
+            /// </summary>
+            PerformAutomaticRecovery
         }
 
-        public int RemotePort
+
+        private enum RecoveryConnectionState
         {
-            get { return m_delegate.RemotePort; }
+            /// <summary>
+            /// Underlying connection is open.
+            /// </summary>
+            Connected,
+            /// <summary>
+            /// In the process of recovering underlying connection.
+            /// </summary>
+            Recovering
         }
+        
+        
+        private Thread m_recoveryThread;
+        private RecoveryConnectionState m_recoveryLoopState = RecoveryConnectionState.Connected;
+
+        private readonly BlockingCollection<RecoveryCommand> m_recoveryLoopCommandQueue = new BlockingCollection<RecoveryCommand>();
+        private readonly CancellationTokenSource m_recoveryCancellationToken = new CancellationTokenSource();
+        private readonly TaskCompletionSource<int> m_recoveryLoopComplete = new TaskCompletionSource<int>();
 
-        public IDictionary<string, object> ServerProperties
+        /// <summary>
+        /// This is the main loop for the auto-recovery thread.
+        /// </summary>
+        private void MainRecoveryLoop()
```

Comment:
This is the core change. Instead of kicking off a Task on the shutdown event this version of AutoRecoveringConnection queues a message in the m_recoveryLoopCommandQueue to switch to auto-recovery mode. While it auto-recovery mode it queues a PerformAutomaticRecovery message at the specified NetworkRecoveryInterval. When the recovery succeeds it goes back into the connected state and the thread sits idle. Closing the connection uses the cancellation token to cancel this recovery loop, and waits on the Complete event to be set so we don't Close on a half-finished recovery (which may not be a problem but seems messy).